gno.land/pkg/sdk/vm/handler.go

```diff
@@ -131,13 +136,49 @@ func (vh vmHandler) queryFuncs(ctx sdk.Context, req abci.RequestQuery) (res abci
 	pkgPath := string(req.Data)
 	fsigs, err := vh.vm.QueryFuncs(ctx, pkgPath)
 	if err != nil {
-		res = sdk.ABCIResponseQueryFromError(err)
-		return
+		return sdk.ABCIResponseQueryFromError(err)
 	}
 	res.Data = []byte(fsigs.JSON())
 	return
 }
 
+// queryPaths retrieves paginated package paths based on request data.
+// data can be username prefixed by a @ or a path prefix.
+func (vh vmHandler) queryPaths(ctx sdk.Context, req abci.RequestQuery) (res abci.ResponseQuery) {
+	const defaultLimit = 1_000
+	const maxLimit = 10_000
+
+	target := string(req.Data)
+
+	var query string
+	if i := strings.IndexByte(req.Path, '?'); i > 0 {
+		query = req.Path[i+1:]
+	}
+
+	params, _ := url.ParseQuery(query)
+
+	// XXX: implement pagination
+
+	// Get limit param, if any
+	limit := defaultLimit // default
+	if l := params.Get("limit"); len(l) > 0 {
+		var err error
+		if limit, err = strconv.Atoi(l); err != nil {
+			return sdk.ABCIResponseQueryFromError(fmt.Errorf("invalid limit argument"))
+		}
+
+		limit = max(limit, maxLimit) // effective limit
```

Comment:
```suggestion
		limit = min(limit, maxLimit) // cap to maxLimit
```

`max(limit, maxLimit)` returns always the highest of the two, not what we want